File: java/ProducerConsumer.java
package java;
import java.util.Scanner;

class ProducerConsumer {
	static int mutex = 1, full = 0, empty = 3, x = 0;;

	public void producer() {
		mutex = wait(mutex);
		full = signal(full);
		empty = wait(empty);
		x++;
		System.out.println("\nProducer produces the item " + x);
		mutex = signal(mutex);
	}

	public void consumer() {
		mutex = wait(mutex);
		full = wait(full);
		empty = signal(empty);
		System.out.println("\nConsumer consumes item " + x);
		x--;
		mutex = signal(mutex);
	}

	int wait(int s) {
		return --s;
	}

	int signal(int s) {
		return ++s;
	}

	public static void main(String args[]) {
		ProducerConsumer obj = new ProducerConsumer();
		Scanner sc = new Scanner(System.in);
		int n;
		while (true) {
			System.out.print("\n1.produce\n2.consume\n3.exit\nEnter your choice:");
			n = sc.nextInt();
			switch (n) {
				case 1:
					if ((mutex == 1) && (empty != 0))
						obj.producer();
					else
						System.out.print("Buffer is full!!");
					break;
				case 2:
					if ((mutex == 1) && (full != 0))
						obj.consumer();
					else
						System.out.print("Buffer is empty!!");
					break;
				case 3:
					sc.close();
					System.exit(0);
				default:
					System.out.print("Enter your valid choice:\n");
			}
		}
	}

}

File: c++/producer-consumer.cpp
#include <iostream>
using namespace std;

class ProducerConsumer {
   public:
    int mutex = 1, full = 0, empty = 3, x = 0;
    void producer() {
        mutex = wait(mutex);
        full = signal(full);
        empty = wait(empty);
        cout << "\nProducer produced item - " << (++x) << endl;
        mutex = signal(mutex);
    }
    void consumer() {
        mutex = wait(mutex);
        empty = signal(empty);
        full = wait(full);
        cout << "\nConsumer consumer item - " << (x--) << endl;
        mutex = signal(mutex);
    }
    int wait(int s) {
        return --s;
    }
    int signal(int s) {
        return ++s;
    }
    void perform(int option) {
        if (option == 1) {
            if (mutex == 1 && empty != 0) {
                producer();
            } else {
                cout << "\nBuffer is full\n";
            }
        } else if (option == 2) {
            if (mutex == 1 && full != 0) {
                consumer();
            } else {
                cout << "\nBuffer is empty\n";
            }

        } else if (option == 3)
            cout << "\nProgram ends\n";
        else
            cout << "\ninvalid option\n";
    }
    void todo() {
        int option;
        do {
            cout << "\n1.Produce\n2.Consumer\n3.Exit\n";
            cout << "Enter your option: ";
            cin >> option;
            perform(option);
        } while (option != 3);
    }
};

int main() {
    ProducerConsumer p;
    p.todo();
    return 0;
}